projects/igniteui-angular/src/lib/grids/grid/grid.pinning.spec.ts

```diff
@@ -1,1140 +1,681 @@
-            fix.detectChanges();
-
-            // verify column is pinned
-            expect(grid.pinnedColumns.length).toEqual(2);
-            expect(grid.unpinnedColumns.length).toEqual(9);
-
-            // verify container widths
-            expect(grid.pinnedWidth).toEqual(400);
-            expect(grid.unpinnedWidth + grid.scrollWidth).toEqual(400);
-
-            expect(col.pinned).toBe(true);
-            expect(col.visibleIndex).toEqual(1);
-
-            cell = grid.getCellByColumn(0, 'CompanyName');
-            expect(cell.visibleColumnIndex).toEqual(1);
-            expect(cell.nativeElement.classList.contains(FIXED_CELL_CSS)).toBe(true);
-        }));
-
-        it('should allow pinning/unpinning via the column API', fakeAsync(() => {
-            const fix = TestBed.createComponent(DefaultGridComponent);
-            fix.detectChanges();
-            const grid = fix.componentInstance.instance;
-
-            const col = grid.getColumnByName('ID');
-
-            col.pinned = true;
-            tick();
-            fix.detectChanges();
-
-            // verify column is pinned
-            expect(col.pinned).toBe(true);
-            expect(col.visibleIndex).toEqual(2);
-
-            expect(grid.pinnedColumns.length).toEqual(3);
-            expect(grid.unpinnedColumns.length).toEqual(8);
 
-            // verify container widths
-            expect(grid.pinnedWidth).toEqual(600);
-            expect(grid.unpinnedWidth + grid.scrollWidth).toEqual(200);
+        describe('Initially pinned columns', () => {
+            let fix;
+            let grid: IgxGridComponent;
+
+            beforeEach(fakeAsync(() => {
+                fix = TestBed.createComponent(PinOnInitAndSelectionComponent);
+                fix.detectChanges();
+                grid = fix.componentInstance.grid;
+            }));
+
+            it('should correctly initialize when there are initially pinned columns.', () => {
+
+                // verify pinned/unpinned collections
+                expect(grid.pinnedColumns.length).toEqual(2);
+                expect(grid.unpinnedColumns.length).toEqual(9);
+
+                // verify DOM
+                const firstIndexCell = grid.getCellByColumn(0, 'CompanyName');
+                expect(firstIndexCell.visibleColumnIndex).toEqual(0);
+
+                const lastIndexCell = grid.getCellByColumn(0, 'ContactName');
+                expect(lastIndexCell.visibleColumnIndex).toEqual(1);
+                expect(GridFunctions.isCellPinned(lastIndexCell)).toBe(true);
+
+                const headers = GridFunctions.getColumnHeaders(fix);
+
+                expect(headers[0].context.column.field).toEqual('CompanyName');
+
+                expect(headers[1].context.column.field).toEqual('ContactName');
+                expect(headers[1].parent.nativeElement.classList.contains(FIXED_HEADER_CSS)).toBe(true);
+
+                // verify container widths
+                expect(grid.pinnedWidth).toEqual(400);
```

Comment:
It is not good approach to use fixed constants for checking width. You can add this check:
expect(grid.pinnedWidth).toBeGreaterThanOrEqual(399);
expect(grid.pinnedWidth).toBeLessThanOrEqual(401); 
because it is possible to have 1px difference

Reply:
Done